File: python/src/fds/protobuf/stach/extensions/v3/StachV3ArrowHelper.py
import pyarrow as pa
import pyarrow.ipc as ipc
import base64
import pandas as pd
import json
from google.protobuf import json_format
import io

from fds.protobuf.stach.extensions.v3.StachV3Constants import (
    ignored_columns,
    group_prefix,
    level_name_column,
    security_name_column,
    column_separator
)

try:
    from fds.protobuf.stach.table import Table_pb2
    from fds.protobuf.stach.v3 import Views_pb2
except ImportError as e:
    # Handle case where protobuf messages are not available
    Table_pb2 = None
    Views_pb2 = None

def get_data_row(record_batch, row_index, data_table, column_headers_mapping):
    data_row = {}
    security_column = next((k for k, v in column_headers_mapping.items() if security_name_column in v.lower()), None)
    security_name_column_data = record_batch.column(security_column) if security_column else None
    level_column = next((k for k, v in column_headers_mapping.items() if level_name_column in k.lower()), None)
    level_column_data = record_batch.column(level_column) if level_column else None
    level_index = int(get_column_value(level_column_data, row_index)) if level_column_data else 0

    for column in column_headers_mapping:
        if group_prefix not in column:
            column_data = record_batch.column(column) if column else None
            column_name = column_headers_mapping[column]
            if column_data is not None and column_name:
                data_row[column_name] = get_column_value(column_data, row_index)
        else:
            column_name = column
            group_index = int(column_name.replace(group_prefix, ""))
            if level_index == group_index:
                if security_name_column_data is not None:
                    data_row[column_name] = get_column_value(security_name_column_data, row_index)
                else:
                    data_row[column_name] = ''
            else:
                if row_index > 0 and level_index > 0 and group_index < level_index:
                    if security_name_column_data is not None:
                        data_row[column_name] = data_table.iloc[row_index - 1][column_name]
                    else:
                        data_row[column_name] = ''
    return data_row

def is_arrow_file(stream):
    arrow_file_magic = b'ARROW1'
    stream.seek(0)
    buffer = stream.read(6)
    stream.seek(0)
    return buffer == arrow_file_magic

def get_arrow_stream(arrow_bytes):    
    arrow_stream = io.BytesIO(arrow_bytes)
    if is_arrow_file(arrow_stream):
        arrow_stream.seek(0)
        return ipc.RecordBatchFileReader(arrow_stream)
    else:
        arrow_stream.seek(0)
        return ipc.RecordBatchStreamReader(arrow_stream)

def get_levels(arrow_bytes: bytes):
    levels = []
    reader = get_arrow_stream(arrow_bytes)
    if reader is None:
        raise Exception("Failed to create ArrowStreamReader from arrow_bytes.")
    
     # Read first record batch
    try:
       record_batch = reader.read_next_batch()
    except AttributeError:
        # For pa.ipc.RecordBatchFileReader, use get_batch(0)
        if hasattr(reader, "num_record_batches") and reader.num_record_batches > 0:
            record_batch = reader.get_batch(0)
        else:
            record_batch = None

    if record_batch is not None:
         try:
             column_index = record_batch.schema.get_field_index("level")
         except (KeyError, ValueError):
             column_index = -1
         if column_index >= 0:
             column = record_batch.column(column_index)
             levels = [int(x.as_py()) for x in column]
    return levels
    
def get_column_value(column_array, row_index):    
    if column_array is None:
        return ''
    if pa.types.is_int32(column_array.type):
        return str(column_array[row_index].as_py())
    elif pa.types.is_int64(column_array.type):
        return str(column_array[row_index].as_py())
    elif pa.types.is_string(column_array.type):
        return column_array[row_index].as_py()
    elif pa.types.is_boolean(column_array.type):
        return str(column_array[row_index].as_py())
    elif pa.types.is_float64(column_array.type):
        return str(column_array[row_index].as_py())
    else:
        return ''

def read_multi_level_headers_schema(multi_level_headers_records, schema):
    columns = []
    n_columns = len(schema)
    n_batches = len(multi_level_headers_records)
    for i in range(n_columns):
        column_arrays = []
        for j in range(n_batches):
            col_value = multi_level_headers_records[j].column(i)
            column_arrays.append(col_value)
        columns.append(pa.chunked_array(column_arrays))
    return pa.Table.from_arrays(columns, schema.names)

def generate_multi_level_headers(multi_level_headers_array_decoded):
    multi_level_headers_records = []
    mlh_table = None
    try:
        memory_stream = pa.BufferReader(multi_level_headers_array_decoded)
        reader = ipc.RecordBatchFileReader(memory_stream)
        for i in range(reader.num_record_batches):
            rb = reader.get_batch(i)
            multi_level_headers_records.append(rb)
        mlh_table = read_multi_level_headers_schema(multi_level_headers_records, reader.schema)
    except Exception as ex:
        # Try as stream if file fails
        memory_stream = pa.BufferReader(multi_level_headers_array_decoded)
        reader = ipc.RecordBatchStreamReader(memory_stream)
        for rb in reader:
            multi_level_headers_records.append(rb)
        mlh_table = read_multi_level_headers_schema(multi_level_headers_records, reader.schema)
    return mlh_table

def _read_arrow_stream(file_path):
    """Read arrow file and return bytes and stream reader."""
    with open(file_path, "rb") as f:
        arrow_bytes = f.read()

    reader = get_arrow_stream(arrow_bytes)
    if reader is None:
        raise Exception("Failed to create ArrowStreamReader")

    return arrow_bytes, reader

def _setup_level_columns(arrow_bytes, column_headers_mapping):
    """Read levels from arrow data and add group columns to mapping."""
    levels = get_levels(arrow_bytes)
    if levels:
        max_level = max(levels)
        for i in range(max_level + 1):
            level_column_name = f"{group_prefix}{i}"
            if level_column_name not in column_headers_mapping:
                column_headers_mapping[level_column_name] = level_column_name

def _parse_table_metadata(metadata):
    """Parse and return Table protobuf object from metadata."""
    stach_table = None
    if Table_pb2 is None:
        return stach_table
    if "fds:stach:table" in metadata:
        try:
            table_object = metadata["fds:stach:table"]
            table_array_decoded = base64.b64decode(table_object)
            stach_table = Table_pb2.Table()
            stach_table.ParseFromString(table_array_decoded)
        except Exception as e:
            # Handle parsing errors gracefully - continue without table metadata
            pass
    return stach_table

def _parse_views_metadata(metadata):
    """Parse and return view headers from metadata."""
    view_headers = {}
    if Views_pb2 is None:
        return view_headers
    if "fds:stach:views" in metadata:
        try:
            view_object = metadata["fds:stach:views"]
            views_array_decoded = base64.b64decode(view_object)
            views = Views_pb2.Views()
            views.ParseFromString(views_array_decoded)
            if views.views:
                first_view = views.views[0]
                if first_view.table.headers:
                    view_headers = dict(first_view.table.headers)
        except Exception as e:
            # Handle parsing errors gracefully - continue without view headers
            pass
    return view_headers

def _parse_multi_level_headers(schema_metadata, stach_table, view_headers):
    """Parse multi-level headers table and return concatenated columns list."""
    concat_columns_list = {}
    if b"fds:stach:multiLevelHeadersTable" in schema_metadata:
        try:
            multi_level_headers_table_object = schema_metadata[b"fds:stach:multiLevelHeadersTable"].decode()
            multi_level_headers_array_decoded = base64.b64decode(multi_level_headers_table_object)
            if multi_level_headers_array_decoded:
                try:
                    multi_level_headers = generate_multi_level_headers(multi_level_headers_array_decoded)

                    if multi_level_headers:
                        for row_index in range(multi_level_headers.num_rows):
                            concat_column_value = ''
                            primary_key_column = ''

                            for col_index in range(multi_level_headers.num_columns):
                                column_object = multi_level_headers.column(col_index)
                                column_name = multi_level_headers.column_names[col_index]
                                header_column_value = get_column_value(column_object, row_index)

                                if column_name == getattr(stach_table.split_result, "multi_level_headers_table_reference", ""):
                                    primary_key_column = header_column_value
                                if header_column_value:
                                    if view_headers and header_column_value in view_headers:
                                        header_column_value = view_headers.get(header_column_value, header_column_value)
                                    concat_column_value = header_column_value + column_separator + concat_column_value

                            if primary_key_column:
                                concat_columns_list[primary_key_column] = concat_column_value.rstrip(column_separator)
                except Exception as exc:
                    # Handle inner parsing errors gracefully
                    pass
        except Exception as e:
            # Handle outer parsing errors gracefully
            pass

    return concat_columns_list

def _build_column_headers_mapping(schema, view_headers, concat_columns_list, column_headers_mapping):
    """Build column headers mapping from schema field names."""
    for field in schema:
        header_name = field.name
        if concat_columns_list:
            header_name = concat_columns_list.get(field.name, field.name)
        else:
            header_name = view_headers.get(field.name, field.name)
        column_headers_mapping[field.name] = header_name

def _populate_dataframe(reader, column_headers_mapping):
    """Create and populate DataFrame from record batches."""
    columns = [column_headers_mapping[k] for k in column_headers_mapping]
    data_table = pd.DataFrame(columns=columns)

    if hasattr(reader, "num_record_batches"):
        for idx in range(reader.num_record_batches):
            batch = reader.get_batch(idx)
            for row_index in range(batch.num_rows):
                data_row = get_data_row(batch, row_index, data_table, column_headers_mapping)
                data_table = pd.concat([data_table, pd.DataFrame([data_row])], ignore_index=True)
    else:
        for record_batch in reader:
            for row_index in range(record_batch.num_rows):
                data_row = get_data_row(record_batch, row_index, data_table, column_headers_mapping)
                data_table = pd.concat([data_table, pd.DataFrame([data_row])], ignore_index=True)

    return data_table

def _remove_ignored_columns(data_table, ignored_columns):
    """Remove ignored columns from DataFrame."""
    for column in ignored_columns:
        if column in data_table.columns:
            data_table.drop(column, axis=1, inplace=True)
    return data_table

File: dotnet/StachExtensions/FactSet.Protobuf.Stach.Extensions/V3/StachV3ArrowHelper.cs
using Apache.Arrow;
using Apache.Arrow.Ipc;
using System;
using System.Collections.Generic;
using System.Data;
using System.IO;
using System.Linq;
using StachV3 = FactSet.Protobuf.Stach.V3;

namespace FactSet.Protobuf.Stach.Extensions.V3
{
    internal static class StachV3ArrowHelper
    {
        public static void BuildGroupColumns(byte[] arrowBytes, ConversionContext ctx)
        {
            var levels = GetLevels(arrowBytes);
            if (levels?.Length > 0)
            {
                var maxLevel = levels.Max();
                for (int i = 0; i <= maxLevel; i++)
                {
                    var levelColumnName = $"{StachV3Constants.GroupPrefix}{i}";
                    if (!ctx.ColumnHeadersMapping.ContainsKey(levelColumnName))
                        ctx.ColumnHeadersMapping.Add(levelColumnName, levelColumnName);
                }
            }
        }

        public static void ParseTableMetadata(IDictionary<string, string> metadata, ConversionContext ctx)
        {
            if (metadata.TryGetValue(ArrowSchemaMetadataNames.Table, out var tableObject))
            {
                ctx.StachTable = StachV3.Table.Parser.ParseFrom(System.Convert.FromBase64String(tableObject)) ?? new StachV3.Table();
            }
        }

        public static void ParseViewsMetadata(IDictionary<string, string> metadata, ConversionContext ctx)
        {
            if (metadata.TryGetValue(ArrowSchemaMetadataNames.Views, out var viewObject))
            {
                var views = StachV3.Views.Parser.ParseFrom(System.Convert.FromBase64String(viewObject)) ?? new StachV3.Views();
                ctx.ViewHeaders = views?.Views_?.FirstOrDefault()?.Table?.Headers?
                                  .ToDictionary(kvp => kvp.Key, kvp => kvp.Value) ?? new Dictionary<string, string>();
            }
        }

        public static void BuildConcatColumnsList(IDictionary<string, string> metadata, ConversionContext ctx)
        {
            if (!metadata.TryGetValue(ArrowSchemaMetadataNames.MultiLevelHeadersTable, out var multiLevelHeadersTableObject))
                return;

            var multiLevelHeadersArrayDecoded = System.Convert.FromBase64String(multiLevelHeadersTableObject);
            if (multiLevelHeadersArrayDecoded?.Length == 0)
                return;

            try
            {
                var multiLevelHeaders = GenerateMultiLevelHeaders(multiLevelHeadersArrayDecoded);
                if (multiLevelHeaders == null)
                    return;

                var multiLevelRef = ctx.StachTable?.SplitResult?.MultiLevelHeadersTableReference;
                bool hasViewHeaders = ctx.ViewHeaders?.Count > 0;
                var separatorChars = StachV3Constants.ColumnSeparator.ToCharArray(); // cache: avoids allocation per row

                // Pre-resolve column names and arrow arrays once — they are row-invariant
                var mlhColumns = new (string Name, IArrowArray Array)[multiLevelHeaders.ColumnCount];
                for (int i = 0; i < multiLevelHeaders.ColumnCount; i++)
                {
                    var col = multiLevelHeaders.Column(i);
                    mlhColumns[i] = (col.Name, col.Data.ArrowArray(0));
                }

                for (int rowIndex = 0; rowIndex < multiLevelHeaders.RowCount; rowIndex++)
                {
                    var concatColumnValue = string.Empty;
                    var primaryKeyColumn = string.Empty;

                    foreach (var (colName, array) in mlhColumns)
                    {
                        var headerColumnValue = GetColumnValue(array, rowIndex);

                        if (colName == multiLevelRef)
                            primaryKeyColumn = headerColumnValue;

                        if (!string.IsNullOrEmpty(headerColumnValue))
                        {
                            if (hasViewHeaders && ctx.ViewHeaders.TryGetValue(headerColumnValue, out var mapped))
                                headerColumnValue = mapped ?? headerColumnValue;

                            concatColumnValue = headerColumnValue + StachV3Constants.ColumnSeparator + concatColumnValue;
                        }
                    }

                    if (!string.IsNullOrEmpty(primaryKeyColumn))
                        ctx.ConcatColumnsList.Add(primaryKeyColumn, concatColumnValue.TrimEnd(separatorChars));
                }
            }
            catch (Exception)
            {
                // Handle exception
            }
        }

        public static void MapColumnHeaders(IList<Field> fields, ConversionContext ctx)
        {
            bool hasConcatColumns = ctx.ConcatColumnsList.Count > 0;
            bool hasViewHeaders = ctx.ViewHeaders.Count > 0;
            foreach (var field in fields)
            {
                string headerName = field.Name;
                if (hasConcatColumns && ctx.ConcatColumnsList.TryGetValue(field.Name, out var concatName))
                    headerName = concatName;
                else if (hasViewHeaders && ctx.ViewHeaders.TryGetValue(field.Name, out var viewName))
                    headerName = viewName;

                ctx.ColumnHeadersMapping.Add(field.Name, headerName);
            }
        }

        public static void BuildDataTableColumns(ConversionContext ctx)
        {
            ctx.DataTable.Columns.AddRange(
                ctx.ColumnHeadersMapping.Select(kvp => new DataColumn(kvp.Value, typeof(string))).ToArray()
            );
        }

        public static void PopulateDataTable(IArrowArrayStream reader, ConversionContext ctx)
        {
            // Pre-compute once: security and level column keys (avoid O(n) search per row)
            var securityColumnKey = ctx.ColumnHeadersMapping
                .FirstOrDefault(x => x.Value != null && x.Value.IndexOf(StachV3Constants.SecurityNameColumn, StringComparison.OrdinalIgnoreCase) >= 0).Key;
            var levelColumnKey = ctx.ColumnHeadersMapping
                .FirstOrDefault(x => x.Key != null && x.Key.IndexOf(StachV3Constants.LevelNameColumn, StringComparison.OrdinalIgnoreCase) >= 0).Key;

            // Pre-split into group columns (with pre-parsed index) and data columns.
            // Avoids repeated groupPrefix checks and int.Parse on every row.
            var groupColumnEntries = ctx.ColumnHeadersMapping
                .Where(kvp => kvp.Key.StartsWith(StachV3Constants.GroupPrefix))
                .Select(kvp => (Key: kvp.Key, GroupIndex: int.Parse(kvp.Key.Substring(StachV3Constants.GroupPrefix.Length))))
                .ToList();

            // Array: fixed schema means fixed size; pre-allocated and reused across batches
            var dataColumnEntries = ctx.ColumnHeadersMapping
                .Where(kvp => !kvp.Key.StartsWith(StachV3Constants.GroupPrefix) && !string.IsNullOrEmpty(kvp.Key) && !string.IsNullOrEmpty(kvp.Value))
                .Select(kvp => (Key: kvp.Key, Name: kvp.Value))
                .ToArray();

            var resolvedDataColumns = new (string Name, IArrowArray Data)[dataColumnEntries.Length];

            // BeginLoadData suspends index/constraint updates during bulk insert;
            // EndLoadData rebuilds them once at the end
            ctx.DataTable.BeginLoadData();
            try
            {
                RecordBatch recordBatch;
                while ((recordBatch = reader.ReadNextRecordBatchAsync().Result) != null)
                {
                    // Resolve column arrays once per batch — column data is stable within a batch
                    IArrowArray securityNameColumnData = !string.IsNullOrEmpty(securityColumnKey) && IsColumnExist(recordBatch, securityColumnKey)
                        ? recordBatch.Column(securityColumnKey)
                        : null;

                    // Cast directly to Int64Array — avoids GetColumnValue string round-trip on every level read
                    Int64Array levelColumnAsInt64 = !string.IsNullOrEmpty(levelColumnKey) && IsColumnExist(recordBatch, levelColumnKey)
                        ? recordBatch.Column(levelColumnKey) as Int64Array
                        : null;

                    // Fill pre-allocated array in-place — no new list or LINQ per batch
                    for (int i = 0; i < dataColumnEntries.Length; i++)
                        resolvedDataColumns[i] = (dataColumnEntries[i].Name, recordBatch.Column(dataColumnEntries[i].Key));

                    for (int rowIndex = 0; rowIndex < recordBatch.Length; rowIndex++)
                    {
                        var dataRow = ctx.DataTable.NewRow();

                        int levelIndex = levelColumnAsInt64 != null
                            ? (int)(levelColumnAsInt64.GetValue(rowIndex) ?? 0L)
                            : 0;

                        // fill data columns
                        for (int i = 0; i < resolvedDataColumns.Length; i++)
                        {
                            var (name, data) = resolvedDataColumns[i];
                            if (data != null)
                                dataRow[name] = GetColumnValue(data, rowIndex);
                        }

                        // fill group columns
                        foreach (var (colKey, groupIndex) in groupColumnEntries)
                        {
                            if (levelIndex == groupIndex)
                            {
                                dataRow[colKey] = securityNameColumnData != null
                                    ? GetColumnValue(securityNameColumnData, rowIndex)
                                    : string.Empty;
                            }
                            else if (rowIndex > 0 && levelIndex > 0 && groupIndex < levelIndex)
                            {
                                dataRow[colKey] = securityNameColumnData != null
                                    ? ctx.DataTable.Rows[rowIndex - 1][colKey]
                                    : (object)string.Empty;
                            }
                        }

                        ctx.DataTable.Rows.Add(dataRow);
                    }
                }
            }
            finally
            {
                ctx.DataTable.EndLoadData();
            }
        }

        public static void RemoveIgnoredColumns(ConversionContext ctx)
        {
            if (StachV3Constants.IgnoredColumns?.Count > 0)
            {
                foreach (var column in StachV3Constants.IgnoredColumns)
                {
                    if (ctx.DataTable.Columns.Contains(column))
                        ctx.DataTable.Columns.Remove(column);
                }
            }
        }

        public static Apache.Arrow.Table GenerateMultiLevelHeaders(byte[] multiLevelHeadersArrayDecoded)
        {
            var multiLevelHeadersRecords = new List<RecordBatch>();
            Apache.Arrow.Table mlhTable = null;

            try
            {
                using (MemoryStream memoryStream = new MemoryStream(multiLevelHeadersArrayDecoded))
                using (var arrowFileReader = new ArrowFileReader(memoryStream, true))
                {
                    var rb = arrowFileReader.ReadNextRecordBatchAsync().Result;
                    while (rb != null)
                    {
                        multiLevelHeadersRecords.Add(rb);
                        rb = arrowFileReader.ReadNextRecordBatchAsync().Result;
                    }

                    mlhTable = ReadMultiLevelHeadersSchema(multiLevelHeadersRecords, arrowFileReader.Schema);
                }
            }
            catch (InvalidDataException ex)
            {
                if (ex.Message == "Invalid magic at offset <6>")
                {
                    using (MemoryStream memoryStream = new MemoryStream(multiLevelHeadersArrayDecoded))
                    using (var arrowStreamReader = new ArrowStreamReader(memoryStream, true))
                    {
                        var rb = arrowStreamReader.ReadNextRecordBatchAsync().Result;
                        while (rb != null)
                        {
                            multiLevelHeadersRecords.Add(rb);
                            rb = arrowStreamReader.ReadNextRecordBatchAsync().Result;
                        }

                        mlhTable = ReadMultiLevelHeadersSchema(multiLevelHeadersRecords, arrowStreamReader.Schema);
                    }
                }
            }

            return mlhTable;
        }

        public static Apache.Arrow.Table ReadMultiLevelHeadersSchema(List<RecordBatch> multiLevelHeadersRecords, Schema schema)
        {
            var nBatches = multiLevelHeadersRecords.Count;
            var nColumns = schema.FieldsLookup.Count;
            var columns = new List<Apache.Arrow.Column>(nColumns);
            var columnArrays = new List<Apache.Arrow.Array>(nBatches);

            for (var i = 0; i < nColumns; i++)
            {
                for (var j = 0; j < nBatches; j++)
                {
                    var colValue = ((RecordBatch)multiLevelHeadersRecords[j]).Column(i);
                    columnArrays.Add((Apache.Arrow.Array)colValue);
                }

                columns.Add(new Apache.Arrow.Column(schema.GetFieldByIndex(i), columnArrays));
                columnArrays.Clear();
            }

            return new Apache.Arrow.Table(schema, columns);
        }

        public static string GetColumnValue(IArrowArray columnArray, int rowIndex)
        {
            if (columnArray == null)
                return string.Empty;

            if (columnArray is Int8Array int8Array)
                return int8Array.GetValue(rowIndex)?.ToString();
            else if (columnArray is Int16Array int16Array)
                return int16Array.GetValue(rowIndex)?.ToString();
            else if (columnArray is Int32Array int32Array)
                return int32Array.GetValue(rowIndex)?.ToString();
            else if (columnArray is Int64Array int64Array)
                return int64Array.GetValue(rowIndex)?.ToString();
            else if (columnArray is StringArray stringArray)
                return stringArray.GetString(rowIndex);
            else if (columnArray is FloatArray floatArray)
                return floatArray.GetValue(rowIndex)?.ToString();
            else if (columnArray is DoubleArray doubleArray)
                return doubleArray.GetValue(rowIndex)?.ToString();
            else if (columnArray is Time32Array time32Array)
                return time32Array.GetValue(rowIndex)?.ToString();
            else if (columnArray is Time64Array time64Array)
                return time64Array.GetValue(rowIndex)?.ToString();
            else if (columnArray is Date32Array date32Array)
                return date32Array.GetValue(rowIndex)?.ToString();
            else if (columnArray is Date64Array date64Array)
                return date64Array.GetValue(rowIndex)?.ToString();
            else if (columnArray is BooleanArray boolArray)
                return boolArray.GetValue(rowIndex)?.ToString();
            else
                return string.Empty;
        }

        public static long[] GetLevels(byte[] arrowBytes)
        {
            var levels = new List<long>();
            var arrowStreamReader = GetArrowStream(arrowBytes) as ArrowStreamReader;
            if (arrowStreamReader == null)
                throw new InvalidOperationException("Failed to create ArrowStreamReader from arrowBytes.");

            var recordBatch = arrowStreamReader.ReadNextRecordBatchAsync().Result;
            if (recordBatch != null)
            {
                int columnIndex = recordBatch.Schema.GetFieldIndex("level");
                if (columnIndex >= 0)
                {
                    var columnX = (Int64Array)recordBatch.Column(columnIndex);
                    for (int i = 0; i < columnX.Values.Length; i++)
                        levels.Add(columnX.Values[i]);
                }
            }
            return levels.ToArray();
        }

        public static bool IsArrowFile(Stream stream)
        {
            byte[] arrowFileMagic = new byte[] { (byte)'A', (byte)'R', (byte)'R', (byte)'O', (byte)'W', (byte)'1' };
            byte[] buffer = new byte[6];
            long originalPosition = stream.Position;
            stream.Seek(0, SeekOrigin.Begin);
            stream.Read(buffer, 0, 6);
            stream.Seek(originalPosition, SeekOrigin.Begin);
            return buffer.SequenceEqual(arrowFileMagic);
        }

        public static IArrowArrayStream GetArrowStream(byte[] arrowBytes)
        {
            if (arrowBytes == null)
                return null;

            // Do NOT use a 'using' here — the MemoryStream must stay alive for as long as
            // the returned reader is in use. The caller is responsible for disposing the reader,
            // which in turn disposes the underlying stream.
            var arrowStream = new MemoryStream(arrowBytes);
            return IsArrowFile(arrowStream)
                ? (IArrowArrayStream)new ArrowFileReader(arrowStream)
                : new ArrowStreamReader(arrowStream);
        }
        
        public static bool IsColumnExist(RecordBatch recordBatch, string columnName)
        {
            if (recordBatch != null)
            {
                int columnIndex = recordBatch.Schema.GetFieldIndex(columnName);
                if (columnIndex >= 0)
                    return true;
            }
            return false;
        }

    }

    // Holds all mutable state for a single conversion call.
    // Using a context object prevents state pollution between concurrent calls
    // and avoids shared static fields.
    public sealed class ConversionContext
    {
        public DataTable DataTable { get; } = new DataTable();
        public Dictionary<string, string> ColumnHeadersMapping { get; } = new Dictionary<string, string>();
        public Dictionary<string, string> ViewHeaders { get; set; } = new Dictionary<string, string>();
        public StachV3.Table StachTable { get; set; } = new StachV3.Table();
        public Dictionary<string, string> ConcatColumnsList { get; } = new Dictionary<string, string>();
    }
}
